_posts/2016/2016-09-08-kubernetes-preparing-nodes.md

```diff
@@ -0,0 +1,259 @@
+FLANNELD_ETCD_ENDPOINTS=https://192.168.0.1:2379,https://192.168.0.2:2379,https://192.168.0.3:2379
+FLANNELD_ETCD_CAFILE=/etc/ssl/etcd/ca.crt
+FLANNELD_ETCD_CERTFILE=/etc/ssl/etcd/etcd-client.crt
+FLANNELD_ETCD_KEYFILE=/etc/ssl/etcd/etcd-client.key
+FLANNELD_IFACE=$PRIMARY_HOST_IP
+FLANNELD_PUBLIC_IP=$PRIMARY_HOST_IP
+EOF
+root@kubernetes:/tmp$ cat > /etc/systemd/system/flanneld.service << EOF
+[Unit]
+Description=Network fabric for containers
+Documentation=https://github.com/coreos/flannel
+Requires=networking.service
+Before=docker.service
+After=networking.service
+
+[Service]
+Type=notify
+Restart=always
+RestartSec=5
+EnvironmentFile=/etc/flanneld/options.env
+LimitNOFILE=40000
+LimitNPROC=1048576
+ExecStartPre=/sbin/modprobe ip_tables
+ExecStartPre=/bin/mkdir -p /run/flanneld
+
+ExecStart=/usr/local/bin/flanneld --ip-masq=true
+
+# Update docker options
+ExecStartPost=/opt/flanneld/mk-docker-opts.sh -d /run/flanneld/docker_opts.env -i
+
+[Install]
+WantedBy=multi-user.target
+EOF
+root@kubernetes:/tmp$ systemctl enable flanneld
+root@kubernetes:/tmp$ systemctl start flanneld
+root@kubernetes:/tmp$ rm -rf flannel-0.5.5
+root@kubernetes:/tmp$ rm -f flannel-0.5.5-linux-amd64.tar.gz
+{% endhighlight %}
+
+## Docker engine
+
+As you can see in _/etc/systemd/system/flanneld.service_ the script _mk-docker-opts.sh_ creates environment variables with the file _/run/flanneld/docker_opts.env_.
+These variables are used by the Docker daemon.
+We use the [drop-in feature](https://www.freedesktop.org/software/systemd/man/systemd.unit.html) of *Systemd*.
+So there is no need to change the standard service definition of Docker.
+The [Installation of the Docker engine](https://docs.docker.com/engine/installation/linux/debian/) is done in the standard way.
+
+{% highlight bash %}
+root@kubernetes:~$ apt-get purge lxc-docker*
+root@kubernetes:~$ apt-get purge docker.io*
+root@kubernetes:~$ apt-get install -y apt-transport-https ca-certificates
+root@kubernetes:~$ apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys 58118E89F3A912897C070ADBF76221572C52609D
+root@kubernetes:~$ cat > /etc/apt/sources.list.d/docker.list << EOF
+deb https://apt.dockerproject.org/repo debian-jessie main
+EOF
+root@kubernetes:~$ apt-get update
+root@kubernetes:~$ apt-get install -y docker-engine
+root@kubernetes:~$ systemctl enable docker
+root@kubernetes:~$ systemctl stop docker
+root@kubernetes:~$ mkdir -p /etc/systemd/system/docker.service.d
+root@kubernetes:~$ cat > /etc/systemd/system/docker.service.d/docker.conf << "EOF"
+[Service]
+EnvironmentFile=/run/flanneld/docker_opts.env
+ExecStart=
+ExecStart=/usr/bin/docker daemon -H fd:// $DOCKER_OPTS $DOCKER_OPT_BIP $DOCKER_OPT_MTU $DOCKER_OPT_IPMASQ
+EOF
+root@kubernetes:~$ systemctl daemon-reload
+root@kubernetes:~$ systemctl enable docker
+root@kubernetes:~$ systemctl restart docker
+{% endhighlight %}
+
+*drop-in* overwrites the start command.
+The new one which is loaded via the option *EnvironmentFile*.
```

Comment:
The new one contains the variables from the file _/run/flanneld/docker_opts.env_ generated by the flanneld and loaded via the option *EnvironmentFile*.